pyresample/bilinear/__init__.py

```diff
@@ -0,0 +1,518 @@
+                                   nprocs=nprocs)
+
+    del output_idxs, dists
+
+    # Reduce index reference
+    input_size = input_idxs.sum()
+    index_mask = (idx_ref == input_size)
+    idx_ref = np.where(index_mask, 0, idx_ref)
+
+    # Get output projection as pyproj object
+    proj = Proj(out_area.proj4_string)
+
+    # Get output x/y coordinates
+    out_x, out_y = _get_output_xy(out_area, proj)
+
+    # Get input x/ycoordinates
+    in_x, in_y = _get_input_xy(in_area, proj, input_idxs, idx_ref)
+
+    # Get the four closest corner points around each output location
+    pt_1, pt_2, pt_3, pt_4, idx_ref = \
+        _get_bounding_corners(in_x, in_y, out_x, out_y, neighbours, idx_ref)
+
+    # Calculate vertical and horizontal fractional distances t and s
+    t__, s__ = _get_ts(pt_1, pt_2, pt_3, pt_4, out_x, out_y)
+
+    # Remove mask and put np.nan at the masked locations instead
+    if masked:
+        mask = np.isnan(t__) | np.isnan(s__)
+        t__ = np.ma.masked_where(mask, t__)
+        s__ = np.ma.masked_where(mask, s__)
+
+    return t__, s__, input_idxs, idx_ref
+
+
+def _get_ts(pt_1, pt_2, pt_3, pt_4, out_x, out_y):
+    """Calculate vertical and horizontal fractional distances t and s"""
+
+    t__ = np.nan * np.zeros(out_y.shape)
+    s__ = np.nan * np.zeros(out_y.shape)
+
+    vert_parallel = _find_vert_parallels(pt_1, pt_2, pt_3, pt_4)
+    horiz_parallel = _find_horiz_parallels(pt_1, pt_2, pt_3, pt_4)
+
+    # Cases where verticals are parallel
+    idxs = vert_parallel & np.invert(horiz_parallel)
+    if np.any(idxs):
+        t__[idxs], s__[idxs] = \
+            _get_ts_uprights_parallel(pt_1[idxs, :], pt_2[idxs, :],
+                                      pt_3[idxs, :], pt_4[idxs, :],
+                                      out_y[idxs, :], out_x[idxs, :])
+
+    # Cases where both verticals and horizontals are parallel
+    idxs = vert_parallel & horiz_parallel
+    if np.any(idxs):
+        t__[idxs], s__[idxs] = \
+            _get_ts_parallellogram(pt_1[idxs, :], pt_2[idxs, :], pt_3[idxs, :],
+                                   out_y[idxs, :], out_x[idxs, :])
+
+    # All the rest, ie. where the verticals are not parallel
+    idxs = np.invert(vert_parallel)
+    if np.any(idxs):
+        t__[idxs], s__[idxs] = \
+            _get_ts_irregular(pt_1[idxs, :], pt_2[idxs, :],
+                              pt_3[idxs, :], pt_4[idxs, :],
+                              out_y[idxs], out_x[idxs])
+
+    with np.errstate(invalid='ignore'):
+        idxs = (t__ < 0) | (t__ > 1) | (s__ < 0) | (s__ > 1)
+    t__[idxs] = np.nan
+    s__[idxs] = np.nan
+
+    return t__, s__
+
+
+def _find_vert_parallels(pt_1, pt_2, pt_3, pt_4):
+    """Find vertical parallels from rectangle defined by four (x, y)
+    points"""
+    vals = (pt_3[:, 0] - pt_1[:, 0]) * (pt_4[:, 1] - pt_2[:, 1]) - \
+           (pt_4[:, 0] - pt_2[:, 0]) * (pt_3[:, 1] - pt_1[:, 1])
+    idxs = vals == 0.0
```

Comment:
That wasn't mentioned in the originating paper, but you could be right that machine epsilon can cause problems in some corner cases.